### ISSUE TITLE
Support multi-document editing via browser tabs

### ISSUE THREAD

**Issue by davidje13** [davidje13/SequenceDiagram#60]
This is a prerequisite to #59

When multiple tabs are open, each should be able to track its own state. This can be achieved by using unique URLs negotiated between the tabs and storing data for each URL individually. If any tab is bookmarked it will reopen using its own previously captured code. If a tab is duplicated, one of the resulting tabs will spawn a copy document using a new ID.

It may be desirable to store timestamps with each document and discard the oldest if local storage space becomes an issue.

Negotiation between tabs can be achieved using local `StorageEvent` listeners (`BroadcastChannel` is more appropriate but not supported by Safari).

URLs should be denoted using client-side identifiers to encourage caching of resources (i.e. `#1`, `#2`, etc.)

When visiting the root page when multiple documents are available, it may be desirable to offer a choice of document to open. It may also be desirable to offer a way of clearing documents from the browser.

**Comment by davidje13** [comment]
Envisioned flow:

1. The user visits `/`. No documents exist in memory so document 1 is created and the URL is rewritten to `/#1`.
2. The user edits the document; it is saved to slot 1.
3. The user closes the document
4. The user visits `/`. Only one document exists so the URL is rewritten to `/#1` and the existing document loaded. A marker entry is written to broadcast that an existing document was loaded from slot 1. No tabs are listening for this so nothing further happens.
5. In a new tab, the user visits `/#1`. The existing document is loaded and a marker entry is written. The old tab is notified about this change and responds with a new marker entry. The new tab receives this notification and rewrites its URL to `/#2`, storing a copy of slot 1's document in slot 2.

The values of marker entries are not important; they must differ from the old value (if any) but otherwise can be anything. It is the change notification which matters.

I'm not sure whether storage notifications are guaranteed to reach background tabs (especially if they are old and the browser has chosen to hibernate them), so there could be some quirks. It may be worth adding additional copy-on-write behaviour (check if old value matches expectation and if not, clone the document before writing). This is more reliable but would not be a good solution on its own, as the URL would only change once an edit is made.

**Comment by davidje13** [comment]
This has been implemented, but turned out more complicated than expected due to some quirks in how CodeMirror responds to duplicating tabs (looks like its internal text buffer gets "restored" by the browser after duplicating, which confuses it, causing the contents of the text editor to be clobbered with recent text). The workaround is a bit involved but should be safe (will not clobber old document data, but might cause newly cloned documents to be incorrect in some situations).

The UI is minimal but functional; I'm not entirely sure about some of the choices I made here but it can be iterated on (e.g. it may not be intuitive to users that opening the same document in multiple tabs will clone it, especially if they have forgotten they have it open in an old tab). Tested in Chrome, FireFox and Safari.
